scrapy/core/downloader/__init__.py

```diff
@@ -99,28 +98,33 @@ def _deactivate(response):
     def needs_backout(self):
         return len(self.active) >= self.total_concurrency
 
-    def _get_slot(self, request, spider):
-        key = self._get_slot_key(request, spider)
-        if key not in self.slots:
-            conc = self.ip_concurrency if self.ip_concurrency else self.domain_concurrency
-            conc, delay = _get_concurrency_delay(conc, spider, self.settings)
-            self.slots[key] = Slot(conc, delay, self.randomize_delay)
-
-        return key, self.slots[key]
+    def _enqueue_request(self, request, spider):
+        d = defer.maybeDeferred(self._get_slot_key, request)
+        d.addCallback(self._slot_request, request, spider)
+        return d
 
-    def _get_slot_key(self, request, spider):
+    def _get_slot_key(self, request):
         if self.DOWNLOAD_SLOT in request.meta:
             return request.meta[self.DOWNLOAD_SLOT]
-
         key = urlparse_cached(request).hostname or ''
         if self.ip_concurrency:
-            key = dnscache.get(key, key)
-
+            d = reactor.resolve(key)
+            def fallback(failure):
```

Comment:
Do you know what happens currently when resolution fails? Are we going to shallow some errors, which were popping up before?

Reply:
I also wonder how it works with custom download handlers, which don't necessarily use http. 
E.g. you have a file URL, or data URL, or s3 url - is behavior changing for them?

Reply:
Previously if there was nothing in cache, original key was used; now it is a fixed string.

Reply:
> Are we going to shallow some errors…

Should we log the failure?

> …which were popping up before?

I assume that a request to the same domain after a DNS error will retry a DNS resolution. Should I look for a way to test this?

> I also wonder how it works with custom download handlers, which don't necessarily use http.
E.g. you have a file URL, or data URL, or s3 url - is behavior changing for them?
> Previously if there was nothing in cache, original key was used; now it is a fixed string.

Good call.

I’m thinking we could add a `dns_resolvable` flag to download handler classes (`False` by default, safer for user-made download handlers), implement `DownloadHandlers.is_dns_resolvable(request)`, and use the old behavior where DNS resolution is not possible.

I assume only `datauri` and `file` would have it set to `False`. For other requests, the behavior would indeed change, but it should be to the better.